src/loan.rs

```diff
@@ -60,21 +60,177 @@ static COVENANT_PK: &str = "03b9b6059008e3576aad58e05a3a3e37133b05f68cda8535ec09
 /// `covenant_pk`, which is only used to verify that the transaction
 /// data on the witness stack matches the transaction which triggered
 /// the call.
-fn collateral_descriptor(
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct CollateralContract {
+    descriptor: Descriptor<PublicKey>,
     borrower_pk: PublicKey,
     lender_pk: PublicKey,
-    timelock: u64,
-    repayment_output: TxOut,
-) -> Result<Descriptor<PublicKey>> {
-    let repayment_output = serialize(&repayment_output).to_hex();
-    let desc = Descriptor::<elements::bitcoin::PublicKey>::from_str(
-        &(format!(
-            "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
-            COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
-        )),
-    )?;
-
-    Ok(desc)
+    timelock: u32, // TODO: Model timelocks as u32
+}
+
+impl CollateralContract {
+    /// Fill in the collateral contract template with the provided arguments.
+    fn new(
+        borrower_pk: PublicKey,
+        lender_pk: PublicKey,
+        timelock: u64,
+        repayment_output: TxOut,
+    ) -> Result<Self> {
+        let repayment_output = serialize(&repayment_output).to_hex();
+        let descriptor = Descriptor::<elements::bitcoin::PublicKey>::from_str(
+            &(format!(
+                "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
+                COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
+            )),
+        )
```

Comment:
If you want, this could be replaced with `.parse` (assuming `from_str` is the `FromStr` implementation) and not just an associated function.

Reply:
That would cut some of the noise because type-inference would do the rest.

Reply:
I always forget about `.parse`, thank you!

Reply:
Done here: https://github.com/comit-network/baru/commit/127b5d825f9fde75ca5d4f34b72aaf46ebcaa20a